### PR TITLE
WTH - Additional details

### PR DESCRIPTION
**Pull request by wtholt**
Created new Additional Details module where Service Providers can create service specific surveys in order to gain feedback from users in SPARC-Request. 

Users can create surveys through Catalog Manager where a user can click 'Add Additional Details', which redirects you to manage the surveys for specific Services. 

There, a user can utilize the form builder tool to create surveys that accept multiple different types of questions. Also, a user can preview how a form would look like before they actually create a new Questionnaire record. 

Once a Questionnaire is created, a user can then manage its 'active' state. Only 'active' questionnaires will be available to take in SPARC-Request. The link to take them will appear on the Document Management page when submitting a new Service Request.

A user does not have to complete the Questionnaire at that time, as they can also take them via links in their user dashboard.

Pivotal Tracker Story Ids

Additional Details Part 1 - #118787429
Additional Details Part 2 - #118787985

### CODE REVIEW

**Comment by amcates** [comment]
First glance, this looks awesome!!!!

File: app/models/service.rb [hunk 1]
```diff
@@ -144,6 +146,10 @@ def self.cents_to_dollars cents
     cents.to_i / 100.0
   end
 
+  def has_additional_details
```
Comment (kyle-glick) [inline app/models/service.rb]: Should be a ? method

File: app/models/service_request.rb [hunk 2]
```diff
@@ -493,6 +493,18 @@ def relevant_service_providers_and_super_users
     identities.flatten.uniq
   end
 
+  def service_ids
+    line_items.map(&:service_id)
+  end
+
+  def services_associated_to_sr
+    service_ids.map{ |si| Service.find(si) }
```
Comment (kyle-glick) [inline app/models/service_request.rb]: Service.where(id: service_ids)

File: app/models/service_request.rb [hunk 3]
```diff
@@ -493,6 +493,18 @@ def relevant_service_providers_and_super_users
...
+  end
+
+  def additional_detail_services
+    services_associated_to_sr.select { |s| s.has_additional_details }
```
Comment (kyle-glick) [inline app/models/service_request.rb]: services_associated_to_sr.select(:&has_additional_details?)

File: app/views/additional_details/previews/_modal_partial.html.haml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+.modal-dialog.modal-lg
+  .modal-content
+    .modal-header
+      %button{ class: 'close', data: { dismiss: 'modal' } }
+        %span
+          &times;
+      .modal-title
+        Additional Detail Survey Preview
```
Comment (kyle-glick) [inline app/views/additional_details/previews/_modal_partial.html.haml]: YML

File: app/views/additional_details/previews/_modal_partial.html.haml [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+    .modal-body
+      = render 'additional_details/submissions/form'
+    .modal-footer
+      %button{ class: 'btn btn-default', data: { dismiss: 'modal' } }
+        Close
```
Comment (kyle-glick) [inline app/views/additional_details/previews/_modal_partial.html.haml]: YML

File: app/views/additional_details/questionnaires/_form.html.haml [hunk 6]
```diff
@@ -0,0 +1,60 @@
+.panel.panel-primary
+  .panel-heading
+    Additional Details Form Builder
```
Comment (kyle-glick) [inline app/views/additional_details/questionnaires/_form.html.haml]: YML

File: app/views/additional_details/previews/_modal_partial.html.haml [hunk 7]
```diff
@@ -0,0 +1,13 @@
+.modal-dialog.modal-lg
```
Comment (kyle-glick) [inline app/views/additional_details/previews/_modal_partial.html.haml]: Needs YML, lines 8, 13

File: app/views/additional_details/questionnaires/_form.html.haml [hunk 8]
```diff
@@ -0,0 +1,60 @@
+.panel.panel-primary
```
Comment (kyle-glick) [inline app/views/additional_details/questionnaires/_form.html.haml]: Needs to be YMLized

File: app/views/additional_details/questionnaires/index.html.haml [hunk 9]
```diff
@@ -0,0 +1,49 @@
+.panel.panel-primary
```
Comment (kyle-glick) [inline app/views/additional_details/questionnaires/index.html.haml]: Needs to be YMLized

File: app/views/additional_details/submissions/_form.html.haml [hunk 10]
```diff
@@ -0,0 +1,25 @@
+.panel.panel-primary
```
Comment (kyle-glick) [inline app/views/additional_details/submissions/_form.html.haml]: Needs to be YMLized

File: app/views/additional_details/submissions/form_partials/_yes_no_form_partial.html.haml [hunk 11]
```diff
@@ -0,0 +1,7 @@
+%br
```
Comment (kyle-glick) [inline app/views/additional_details/submissions/form_partials/_yes_no_form_partial.html.haml]: Needs to be YMLized

File: app/views/dashboard/service_requests/_protocol_service_request_show.html.haml [hunk 12]
```diff
@@ -48,3 +48,9 @@
               = t(:dashboard)[:service_requests][:notifications]
             %th.col-md-5{ data: { field: 'actions', align: 'center' }, colspan: 4 }
               = t(:dashboard)[:service_requests][:actions][:header]
+    - if service_request.additional_detail_services.present?
+      - service_request.additional_detail_services.each do |service|
+        - unless Submission.where(service: service, identity: current_identity).present?
+          %div.alert.alert-warning
+            %h5.text-center
+              = link_to "Please complete the required Additional Details for #{service.name}", new_service_additional_details_submission_path(service)
```
Comment (kyle-glick) [inline app/views/dashboard/service_requests/_protocol_service_request_show.html.haml]: Needs to be YMLized

File: app/views/layouts/additional_details.html.haml [hunk 13]
```diff
@@ -0,0 +1,34 @@
+-# Copyright © 2011 MUSC Foundation for Research Development
```
Comment (kyle-glick) [inline app/views/layouts/additional_details.html.haml]: Needs to be YMLized

File: app/views/service_requests/document_management.html.haml [hunk 14]
```diff
@@ -34,6 +34,9 @@
   .grid_9.col-md-9
     .document_upload_instructions= t("documents.upload_instructions_html")
     .document-management-container
+      - @service_request.additional_detail_services.each do |service|
+        %h1
+          = link_to "Take Additional Details Survey for #{service.name}", new_service_additional_details_submission_path(service), target: '_blank'
```
Comment (kyle-glick) [inline app/views/service_requests/document_management.html.haml]: Needs to be YMLized

**Comment by kyle-glick** [comment]
Only suggestions are:
1. Most of the files are missing the copyright statement. Should it be there?
2. Lots of YMLizing needs to be done
3. You made a lot of 1-2 line partials, but do they need to be partials? What was your reasoning for doing it that way?

**Comment by wtholt** [comment]
@kyle-glick 

Thanks for the suggestions. These are really helpful.

The reason I did partials is because they will render based on an Item's type.

    = render "additional_details/submissions/form_partials/#{item.item_type}_form_partial", item: item, qr: qr


So if item.item_type == 'checkbox' for example, the form will render the partial "checkbox_form_partial"

Seemed like the best way to do it programatically as we are dealing with several different types of questions (checkboxes, radios, multiselects')

File: app/views/layouts/additional_details.html.haml [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+!!!
+%html
+  %head
+    %title Additional Details
+    = stylesheet_link_tag "#{CUSTOM_ASSET_PATH}additional_details/application"
+    = javascript_include_tag 'additional_details/application'
+    = javascript_include_tag 'jquery.qtip.min.js'
+    = csrf_meta_tags
+
+  %body{ style: 'background: none;' }
```
Comment (jayhardee9) [inline app/views/layouts/additional_details.html.haml]: Maybe this inline style should go into a stylesheet.

**Comment by jayhardee9** [comment]
Killer PR!

**Comment by jwiel86** [comment]
Looks great!

**Pull request merged**